### DISCUSSION TITLE
(Book 1, section 6.4) I don't understand the reasoning behind surface normals calculations and always being against the ray.

### DISCUSSION THREAD

**Discussion by FranAyrolo** [RayTracing/raytracing.github.io#1377] [General]
I finished book 1 entirely, but (besides some math) there's a part that I didn't understand. In secion 6.4 it says (bold for emphasis):  

> We can set things up so that normals always point “outward” from the surface, or always point against the incident ray. This decision is **determined by whether you want to determine the side of the surface at the time of geometry intersection or at the time of coloring.** In this book we have more material types than we have geometry types, so we'll go for less work and put the determination at geometry time. This is simply a matter of preference, and you'll see both implementations in the literature.  

I don't understand how it would be determined at coloring time. From what I can see, geometry intersection time is basically that, the point at which we calculate whether we hit the sphere or not. But how would that happen during coloring?  
  
Also, maybe the order of choice in the explanation is reversed?  
As in, it first explains how to code it to make 'normals always point out', which means determining which side the ray is on happens during coloring, then Listing 17, then 'normals always against the ray', Listing 18, and then the paragraph I quoted.  
Shouldn't that paragraph read:  
> We can set things up so that normals always point “outward” from the surface, or always point against the incident ray. This decision is determined by whether you want to determine the side of the surface at the time of COLORING or at the time of GEOMETRY INTERSECTION.

**Comment by dimitry-ishenko** [comment]
Take what I say with a grain of salt (I am learning just like you), but my understanding is:

> at the time of geometry intersection

... means inside the `hit()` function of the object where the intersection takes place (in our case inside the `sphere` class).

> at the time of coloring

... means inside the `ray_color()` function of the `camera` class.
